add support of TS `Optional variance annotations`

magic-akari opened the pull request:
This PR add support of TS `Optional variance annotations`
-  https://github.com/microsoft/TypeScript/pull/48240

magic-akari committed:
add support of TS `Optional variance annotations`

magic-akari committed:
chore: add JSX test cases

evanw commented on internal/js_parser/ts_parser_test.go:
```diff
@@ -158,6 +158,29 @@ func TestTSTypes(t *testing.T) {
+	expectPrintedTSX(t, "<in T extends={true}></in>", "/* @__PURE__ */ React.createElement(\"in\", {\n  T: true,\n  extends: true\n});\n")
+	expectPrintedTSX(t, "<out T extends={true}></out>", "/* @__PURE__ */ React.createElement(\"out\", {\n  T: true,\n  extends: true\n});\n")
+	expectPrintedTSX(t, "<in out T extends={true}></in>", "/* @__PURE__ */ React.createElement(\"in\", {\n  out: true,\n  T: true,\n  extends: true\n});\n")
+	expectPrintedTSX(t, "<in T,>() => {}", "() => {\n};\n")
```
Where did these test cases come from? I can't reproduce this. The [latest nightly build of the TypeScript compiler](https://www.typescriptlang.org/play?ts=4.7.0-dev.20220331#code/DwSwdgBAKgNAfACgJQQLxwgbwL5A) gives this for `<in T,>() => {}` instead when JSX is enabled:

```js
React.createElement("in", { T: true }, "() => ");
```

and this when JSX is disabled:

```js
T,  > ();
{ }
```

magic-akari replied:
Oh, thanks.
I will report this issue to TypeScript repo.

magic-akari replied:
You are right.
It's TypeScript Error `TS1273`.

I have modified the code and added some more test cases.

magic-akari committed:
fix: follow TS1273

evanw committed:
Merge branch 'master' into feat/ts-varianceAnnotations

evanw committed:
relocate new test cases

evanw committed:
fix bug with `out` keyword and escapes

evanw committed:
various fixes

- increase test coverage
- fix bug: anonymous classes don't support in/out
- better error messages when in/out isn't allowed
- more robust parsing of invalid in/out sequences

evanw committed:
add a changelog entry
